pkg/snapshot/rule/rule.go

```diff
@@ -0,0 +1,525 @@
+					snap.Metadata.Namespace, snap.Metadata.Name, value)
+				taskTracker := commandTask{}
+
+				err := json.Unmarshal([]byte(value), &taskTracker)
+				if err != nil {
+					err = fmt.Errorf("failed to parse annotation to track running commands on pods due to: %v", err)
+					lastError = err
+					continue
+				}
+
+				if len(taskTracker.pods) == 0 {
+					continue
+				}
+
+				for _, pod := range taskTracker.pods {
+					p, err := k8s.Instance().GetPodByUID(types.UID(pod.uid), pod.namespace)
+					if err != nil {
+						if err == k8s.ErrPodsNotFound {
+							continue
+						}
+
+						logrus.Warnf("failed to get pod with uid: %s due to: %v", pod.uid, err)
+						continue // best effort
+					}
+
+					backgroundPodSet[string(p.GetUID())] = *p
+				}
+
+				err = TerminateCommandInPods(&snap, backgroundPodSet, taskTracker.taskID)
+				if err != nil {
+					err = fmt.Errorf("failed to terminate running commands in pods due to: %v", err)
+					lastError = err
+					continue
+				}
+			}
+		}
+	}
+
+	return lastError
+}
+
+func executeSnapRule(pvcs []v1.PersistentVolumeClaim, snap *crdv1.VolumeSnapshot, rType ruleType) (chan bool, error) {
+	backgroundCommandTermChan := make(chan bool, 1)
+	if snap.Metadata.Annotations != nil {
+		ruleName, present := snap.Metadata.Annotations[ruleAnnotationKeys[rType]]
+		if present && len(ruleName) > 0 {
+			log.SnapshotLog(snap).Infof("Running snap rule: %s", ruleName)
+			taskID, err := uuid.New()
+			if err != nil {
+				err = fmt.Errorf("failed to generate uuid for snapshot rule tasks due to: %v", err)
+				return nil, err
+			}
+
+			pods := make([]v1.Pod, 0)
+			for _, pvc := range pvcs {
+				pvcPods, err := k8s.Instance().GetPodsUsingPVC(pvc.GetName(), pvc.GetNamespace())
+				if err != nil {
+					return nil, err
+				}
+
+				pods = append(pods, pvcPods...)
+			}
+
+			backgroundPodSet := make(map[string]v1.Pod)
+			if len(pods) > 0 {
+				// Get the rule and based on the rule execute it
+				rule, err := k8s.Instance().GetStorkRule(ruleName, snap.Metadata.Namespace)
+				if err != nil {
+					return nil, err
+				}
+
+				cmdExecutorImage := defaultCmdExecutorImage
+				if rule.GetAnnotations() != nil {
+					if imageOverride, ok := rule.GetAnnotations()[cmdExecutorImageOverrideKey]; ok && len(imageOverride) > 0 {
+						cmdExecutorImage = imageOverride
+					}
+				}
+
+				for _, item := range rule.Spec {
+					filteredPods := make([]v1.Pod, 0)
+					// filter pods and only uses the ones that match this selector
+					for _, pod := range pods {
+						if hasSubset(pod.GetObjectMeta().GetLabels(), item.PodSelector) {
+							filteredPods = append(filteredPods, pod)
+						}
+					}
+
+					if len(filteredPods) == 0 {
+						logrus.Warnf("none of the pods matched selectors for rule spec: %v", rule.Spec)
+						continue
+					}
+
+					for _, action := range item.Actions {
+						if action.Type == apis_stork.StorkRuleActionCommand {
+							podsForAction := make([]v1.Pod, 0)
```

Comment:
This function is too big. Move the steps for ActionCommand to a separate function. Other Actions in the future can be separate functions too then.